### PR TITLE
add new metadata types, and correct some minor error

### PR DESCRIPTION
**Pull request by TisserantG**
## What does this PR do ?

 - Create basic documentation for metadata and assetCategory controllers
 - add geopoint type and enum type in metadata (and functionnal tests)
 - correct some minor mistakes

### How should this be manually tested?
You can use Functionnal tests, or call different function of metadata, assetCategory and asset controllers with postman to create asset category and assets using the new metadata types.

### CODE REVIEW

File: features/AssetCategory.feature [hunk 1]
```diff
@@ -284,4 +283,102 @@ Feature: AssetCategory
...
+    When I successfully execute the action "device-manager/metadata":"create" with args:
+      | engineId       | "engine-ayse"          |
+      | body.name      | "color"                |
+      | body.valueType | "string"               |
+      | body.valueList | ["red","blue","green"] |
+      | body.mandatory | true                   |
+    When I successfully execute the action "device-manager/assetCategory":"linkMetadata" with args:
+      | engineId   | "engine-ayse"  |
+      | _id        | "coloredTruck" |
+      | metadataId | "color"        |
```
Comment (Aschen) [inline features/AssetCategory.feature:299]: It feel strange to me to fill the param `metadataId` despite the fact that it was the `name` when is was created

File: features/AssetCategory.feature [hunk 2]
```diff
@@ -284,4 +283,102 @@ Feature: AssetCategory
...
+    When I execute the action "device-manager/assetCategory":"linkMetadata" with args:
+      | engineId   | "engine-ayse"  |
+      | _id        | "coloredTruck" |
+      | metadataId | "color"        |
+      | body.value | "black"        |
+    Then I should receive an error matching:
+      | status | 400 |
```
Comment (Aschen) [inline features/AssetCategory.feature:319]: You should rather test the `error.id` property to ensure it's the correct error who it's returned

File: features/AssetCategory.feature [hunk 3]
```diff
@@ -284,4 +283,102 @@ Feature: AssetCategory
...
+      | body.mandatory | true                   |
+    When I successfully execute the action "device-manager/assetCategory":"linkMetadata" with args:
+      | engineId   | "engine-ayse"  |
+      | _id        | "coloredTruck" |
+      | metadataId | "color"        |
+    When I successfully execute the action "device-manager/assetCategory":"unlinkMetadata" with args:
+      | engineId   | "engine-ayse"  |
+      | _id        | "coloredTruck" |
+      | metadataId | "color"        |
+    When I successfully execute the action "device-manager/assetCategory":"linkMetadata" with args:
```
Comment (Aschen) [inline features/AssetCategory.feature:304]: You could test the content of the document after linking the metadata with the value `red` to ensure everything went well

File: features/AssetCategory.feature [hunk 4]
```diff
@@ -284,4 +283,102 @@ Feature: AssetCategory
...
+      | metadataId | "truckType"        |
+    When I execute the action "device-manager/asset":"create" with args:
+      | engineId                | "engine-ayse" |
+      | body.type               | "truck"       |
+      | body.model              | "M"           |
+      | body.reference          | "asset_02"    |
+      | body.category           | "typedTruck"  |
+      | body.metadata.truckType | "super"       |
+    Then I should receive an error matching:
+      | status | 400 |
```
Comment (Aschen) [inline features/AssetCategory.feature:344]: Same comment about matching the `error.id` property

File: features/AssetCategory.feature [hunk 5]
```diff
@@ -284,4 +283,102 @@ Feature: AssetCategory
...
+    When I successfully execute the action "device-manager/asset":"create" with args:
```
Comment (Aschen) [inline features/AssetCategory.feature:345]: I know this is supposed to fail if something is wrong but you should still get the asset to check if everything went wel

File: lib/controllers/AssetCategoryController.ts [hunk 6]
```diff
@@ -125,7 +125,7 @@ export class AssetCategoryController extends RelationalController {
     const metadataId = request.getString('metadataId');
     const value = request.input.body?.value;
 
-    await this.sdk.document.get(engineId, 'metadata', metadataId); //existance verification
+    const metadataContent = await this.sdk.document.get<MetadataContent>(engineId, 'metadata', metadataId); 
     const category = await this.sdk.document.get<AssetCategoryContent>(engineId, 'asset-category', id);
```
Comment (Aschen) [inline lib/controllers/AssetCategoryController.ts:129]: Those calls can be done in //

File: lib/controllers/AssetController.ts [hunk 7]
```diff
@@ -29,6 +29,11 @@ export class AssetController extends RelationalController {
       description: 'Metadata which are specified in AssetCategory as mandatory must be present',
       message: 'metadata %s is mandatory for the asset',
     });
+    global.app.errors.register('device-manager', 'assetController', 'EnumMetadata', {
```
Comment (Aschen) [inline lib/controllers/AssetController.ts:32]: Error codes should be in camelCase (sorry this isn't documented nor tested in Kuzzle..)
```suggestion
    global.app.errors.register('device-manager', 'asset_controller', 'enum_metadata', {
```

File: lib/core-classes/AssetCategoryService.ts [hunk 8]
```diff
@@ -84,32 +84,48 @@ export class AssetCategoryService {
...
     for (const metadata of metadataList) {
       if (metadata.mandatory) {
-        // eslint-disable-next-line no-prototype-builtins
         if (! (assetMetadata[metadata.name]) && ! this.containsValue(metadataValues, metadata.name)) {
           throw global.app.errors.get('device-manager', 'assetController', 'MandatoryMetadata', metadata.name);
         }
       }
+      if (metadata.valueList) {
+        if (! metadata.valueList.includes(assetMetadata[metadata.name])) {
+          throw global.app.errors.get('device-manager', 'assetController', 'EnumMetadata', metadata.name, assetMetadata[metadata.name]);
```
Comment (Aschen) [inline lib/core-classes/AssetCategoryService.ts:93]:
```suggestion
          throw global.app.errors.get('device-manager', 'asset_controller', 'enum_metadata', metadata.name, assetMetadata[metadata.name]);
```

**Review by Aschen — CHANGES_REQUESTED** [review]

File: features/AssetCategory.feature (hunk 2 above)
Comment (TisserantG) [reply features/AssetCategory.feature:319]: I don't know that, but I'm agree, it's really better, I will do that.

File: features/AssetCategory.feature (hunk 1 above)
Comment (Aschen) [reply features/AssetCategory.feature:299]: If I understand well, you use the metadata name as the document ID right?

If it's the case then my comment does not applies, I thought it was different values. You are right to use the ID to ensure uniqueness :+1:

File: features/AssetCategory.feature (hunk 3 above)
Comment (TisserantG) [reply features/AssetCategory.feature:304]: Done

File: features/AssetCategory.feature (hunk 4 above)
Comment (TisserantG) [reply features/AssetCategory.feature]: Done

File: features/AssetCategory.feature (hunk 5 above)
Comment (TisserantG) [reply features/AssetCategory.feature:349]: Done

File: lib/controllers/AssetCategoryController.ts (hunk 6 above)
Comment (TisserantG) [reply lib/controllers/AssetCategoryController.ts]: Done

File: lib/controllers/AssetController.ts (hunk 7 above)
Comment (TisserantG) [reply lib/controllers/AssetController.ts]: Done (and I correct all the others error code that was present before this PR)

File: lib/core-classes/AssetCategoryService.ts (hunk 8 above)
Comment (TisserantG) [reply lib/core-classes/AssetCategoryService.ts]: done

File: features/AssetCategory.feature (hunk 2 above)
Comment (TisserantG) [reply features/AssetCategory.feature]: done

**Review by Aschen — APPROVED** [review]
LGTM :+1:

**Pull request merged**